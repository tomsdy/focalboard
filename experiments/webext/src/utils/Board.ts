interface BoardFields {
  isTemplate: boolean
}

export default interface Board {
  id: string
  title: string
  fields: BoardFields
}
