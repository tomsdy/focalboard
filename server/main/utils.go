package main

import (
	"crypto/rand"
	"fmt"
	"log"
	"os"
)

// FileExists returns true if a file exists at the path
func fileExists(path string) bool {
	_, err := os.Stat(path)
	if os.IsNotExist(err) {
		return false
	}
	return err == nil
}

// CreateGUID returns a random GUID
func createGUID() string {
	b := make([]byte, 16)
	_, err := rand.Read(b)
	if err != nil {
		log.Fatal(err)
	}
	uuid := fmt.Sprintf("%x-%x-%x-%x-%x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:])

	return uuid
}
