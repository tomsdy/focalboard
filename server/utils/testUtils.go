package utils

import "github.com/stretchr/testify/mock"

var Anything = mock.MatchedBy(func(interface{}) bool { return true })
