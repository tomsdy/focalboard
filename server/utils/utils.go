package utils

import (
	"crypto/rand"
	"fmt"
	"log"
	"time"
)

// CreateGUID returns a random GUID.
func CreateGUID() string {
	b := make([]byte, 16)
	_, err := rand.Read(b)
	if err != nil {
		log.Fatal(err)
	}
	uuid := fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])

	return uuid
}

// GetMillis is a convenience method to get milliseconds since epoch.
func GetMillis() int64 {
	return time.Now().UnixNano() / int64(time.Millisecond)
}
