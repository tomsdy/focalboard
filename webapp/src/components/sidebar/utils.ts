// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {CategoryBoards, DefaultCategory} from '../../store/sidebar'

import {Block} from '../../blocks/block'
import {Board} from '../../blocks/board'

export function addMissingItems(sidebarCategories: Array<CategoryBoards>, allItems: Array<Block | Board>): Array<CategoryBoards> {
    const blocksInCategories = new Map<string, boolean>()
    sidebarCategories.forEach(
        (category) => category.boardIDs.forEach(
            (boardID) => blocksInCategories.set(boardID, true),
        ),
    )

    const defaultCategory: CategoryBoards = {
        ...DefaultCategory,
        boardIDs: [],
    }

    allItems.forEach((block) => {
        if (!blocksInCategories.get(block.id)) {
            defaultCategory.boardIDs.push(block.id)
        }
    })

    // sidebarCategories comes from store,
    // so is frozen and can't be extended.
    // So creating new array from it.
    return [
        ...sidebarCategories,
        defaultCategory,
    ]
}
