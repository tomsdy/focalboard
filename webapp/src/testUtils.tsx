// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'

export const wrapIntl = (children: any) => <IntlProvider locale='en'>{children}</IntlProvider>

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()

        range.getBoundingClientRect = jest.fn()

        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }

        return range
    }
}
