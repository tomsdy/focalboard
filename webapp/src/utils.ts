// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import marked from 'marked'

declare global {
    interface Window {
        msCrypto: Crypto
    }
}

class Utils {
    static createGuid(): string {
        const crypto = window.crypto || window.msCrypto
        function randomDigit() {
            if (crypto && crypto.getRandomValues) {
                const rands = new Uint8Array(1)
                crypto.getRandomValues(rands)
                return (rands[0] % 16).toString(16)
            }

            return (Math.floor((Math.random() * 16))).toString(16)
        }
        return 'xxxxxxxx-xxxx-4xxx-8xxx-xxxxxxxxxxxx'.replace(/x/g, randomDigit)
    }

    static htmlToElement(html: string): HTMLElement {
        const template = document.createElement('template')
        template.innerHTML = html.trim()
        return template.content.firstChild as HTMLElement
    }

    static getElementById(elementId: string): HTMLElement {
        const element = document.getElementById(elementId)
        Utils.assert(element, `getElementById "${elementId}$`)
        return element!
    }

    static htmlEncode(text: string): string {
        return String(text).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;')
    }

    static htmlDecode(text: string): string {
        return String(text).replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/&quot;/g, '"')
    }

    // Markdown

    static htmlFromMarkdown(text: string): string {
        // HACKHACK: Somehow, marked doesn't encode angle brackets
        const renderer = new marked.Renderer()
        renderer.link = (href, title, contents) => `<a target="_blank" rel="noreferrer" href="${href}" title="${title || ''}" onclick="event.stopPropagation();">${contents}</a>`
        const html = marked(text.replace(/</g, '&lt;'), {renderer, breaks: true})
        return html
    }

    // Date and Time

    static displayDate(date: Date): string {
        const dateTimeFormat = new Intl.DateTimeFormat('en', {year: 'numeric', month: 'short', day: '2-digit'})
        const text = dateTimeFormat.format(date)

        return text
    }

    static displayDateTime(date: Date): string {
        const dateTimeFormat = new Intl.DateTimeFormat(
            'en',
            {
                year: 'numeric',
                month: 'short',
                day: '2-digit',
                hour: 'numeric',
                minute: 'numeric',
            })
        const text = dateTimeFormat.format(date)
        return text
    }

    static sleep(miliseconds: number): Promise<void> {
        return new Promise((resolve) => setTimeout(resolve, miliseconds))
    }

    // Errors

    static assertValue(valueObject: any): void {
        const name = Object.keys(valueObject)[0]
        const value = valueObject[name]
        if (!value) {
            Utils.logError(`ASSERT VALUE [${name}]`)
        }
    }

    static assert(condition: any, tag = ''): void {
        /// #!if ENV !== "production"
        if (!condition) {
            Utils.logError(`ASSERT [${tag ?? new Error().stack}]`)
        }

        /// #!endif
    }

    static assertFailure(tag = ''): void {
        /// #!if ENV !== "production"
        Utils.assert(false, tag)

        /// #!endif
    }

    static log(message: string): void {
        /// #!if ENV !== "production"
        const timestamp = (Date.now() / 1000).toFixed(2)
        // eslint-disable-next-line no-console
        console.log(`[${timestamp}] ${message}`)

        /// #!endif
    }

    static logError(message: string): void {
        /// #!if ENV !== "production"
        const timestamp = (Date.now() / 1000).toFixed(2)
        // eslint-disable-next-line no-console
        console.error(`[${timestamp}] ${message}`)

        /// #!endif
    }

    // favicon

    static setFavicon(icon?: string): void {
        const href = icon ? `data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100"><text y=".9em" font-size="90">${icon}</text></svg>` : ''
        const link = (document.querySelector("link[rel*='icon']") || document.createElement('link')) as HTMLLinkElement
        link.type = 'image/x-icon'
        link.rel = 'shortcut icon'
        link.href = href
        document.getElementsByTagName('head')[0].appendChild(link)
    }

    // URL

    static replaceUrlQueryParam(paramName: string, value?: string): void {
        const queryString = new URLSearchParams(window.location.search)
        const currentValue = queryString.get(paramName) || ''
        if (currentValue !== value) {
            const newUrl = new URL(window.location.toString())
            if (value) {
                newUrl.searchParams.set(paramName, value)
            } else {
                newUrl.searchParams.delete(paramName)
            }
            window.history.pushState({}, document.title, newUrl.toString())
        }
    }

    // File names

    static sanitizeFilename(filename: string): string {
        // TODO: Use an industrial-strength sanitizer
        let sanitizedFilename = filename
        const illegalCharacters = ['\\', '/', '?', ':', '<', '>', '*', '|', '"', '.']
        illegalCharacters.forEach((character) => {
            sanitizedFilename = sanitizedFilename.replace(character, '')
        })
        return sanitizedFilename
    }

    // File picker

    static selectLocalFile(onSelect?: (file: File) => void, accept = '.jpg,.jpeg,.png'): void {
        const input = document.createElement('input')
        input.type = 'file'
        input.accept = accept
        input.onchange = async () => {
            const file = input.files![0]
            onSelect?.(file)
        }

        input.style.display = 'none'
        document.body.appendChild(input)
        input.click()

        // TODO: Remove or reuse input
    }

    // Arrays

    static arraysEqual(a: readonly any[], b: readonly any[]): boolean {
        if (a === b) {
            return true
        }
        if (a === null || b === null) {
            return false
        }
        if (a === undefined || b === undefined) {
            return false
        }
        if (a.length !== b.length) {
            return false
        }

        for (let i = 0; i < a.length; ++i) {
            if (a[i] !== b[i]) {
                return false
            }
        }
        return true
    }

    static arrayMove(arr: any[], srcIndex: number, destIndex: number): void {
        arr.splice(destIndex, 0, arr.splice(srcIndex, 1)[0])
    }

    // Clipboard

    static copyTextToClipboard(text: string): boolean {
        const textField = document.createElement('textarea')
        textField.innerText = text
        textField.style.position = 'fixed'
        textField.style.opacity = '0'

        document.body.appendChild(textField)
        textField.select()

        let result = false
        try {
            result = document.execCommand('copy')
        } catch (err) {
            Utils.logError(`copyTextToClipboard ERROR: ${err}`)
            result = false
        }
        textField.remove()

        return result
    }

    static isMobile() {
        const toMatch = [
            /Android/i,
            /webOS/i,
            /iPhone/i,
            /iPad/i,
            /iPod/i,
            /BlackBerry/i,
            /Windows Phone/i,
        ]

        return toMatch.some((toMatchItem) => {
            return navigator.userAgent.match(toMatchItem)
        })
    }
}

export {Utils}
