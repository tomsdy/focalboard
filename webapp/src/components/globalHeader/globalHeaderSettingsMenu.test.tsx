// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'
import {Provider as ReduxProvider} from 'react-redux'

import {render} from '@testing-library/react'

import userEvent from '@testing-library/user-event'
import configureStore from 'redux-mock-store'

import {mocked} from 'ts-jest/utils'

import {wrapIntl} from '../../testUtils'

import TelemetryClient, {TelemetryCategory, TelemetryActions} from '../../telemetry/telemetryClient'

import GlobalHeaderSettingsMenu from './globalHeaderSettingsMenu'

jest.mock('../../telemetry/telemetryClient')
const mockedTelemetry = mocked(TelemetryClient, true)

describe('components/sidebar/GlobalHeaderSettingsMenu', () => {
    const mockStore = configureStore([])
    let store = mockStore({})
    beforeEach(() => {
        store = mockStore({})
    })
    test('settings menu closed should match snapshot', () => {
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <GlobalHeaderSettingsMenu/>
            </ReduxProvider>,
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })

    test('settings menu open should match snapshot', () => {
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <GlobalHeaderSettingsMenu/>
            </ReduxProvider>,
        )

        const {container} = render(component)
        userEvent.click(container.querySelector('.menu-entry') as Element)
        expect(container).toMatchSnapshot()
    })

    test('languages menu open should match snapshot', () => {
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <GlobalHeaderSettingsMenu/>
            </ReduxProvider>,
        )

        const {container} = render(component)
        userEvent.click(container.querySelector('.menu-entry') as Element)
        userEvent.click(container.querySelector('#lang') as Element)
        expect(container).toMatchSnapshot()
    })

    test('imports menu open should match snapshot', () => {
        window.open = jest.fn()
        const component = wrapIntl(
            <ReduxProvider store={store}>
                <GlobalHeaderSettingsMenu/>
            </ReduxProvider>,
        )

        const {container} = render(component)
        userEvent.click(container.querySelector('.menu-entry') as Element)
        userEvent.click(container.querySelector('#import') as Element)
        expect(container).toMatchSnapshot()

        userEvent.click(container.querySelector('[aria-label="Asana"]') as Element)
        expect(mockedTelemetry.trackEvent).toBeCalledWith(TelemetryCategory, TelemetryActions.ImportAsana)
    })
})
