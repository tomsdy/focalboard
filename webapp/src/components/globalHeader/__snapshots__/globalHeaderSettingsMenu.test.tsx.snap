// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/sidebar/GlobalHeaderSettingsMenu imports menu open should match snapshot 1`] = `
<div>
  <div
    class="GlobalHeaderSettingsMenu"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="GlobalHeaderComponent__button menu-entry"
      >
        <i
          class="CompassIcon icon-settings-outline SettingsIcon"
        />
      </div>
      <div
        class="Menu noselect left "
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div>
              <div
                class="MenuOption SubMenuOption menu-option open-left menu-option-active"
                id="import"
              >
                <svg
                  class="SubmenuTriangleIcon Icon"
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="50,35 75,50 50,65"
                  />
                </svg>
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Import
                </div>
                <div
                  class="SubMenu Menu noselect left-bottom"
                >
                  <div
                    class="menu-contents"
                  >
                    <div
                      class="menu-options"
                    >
                      <div
                        aria-label="Import archive"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Import archive
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Trello"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Trello
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Asana"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Asana
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Notion"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Notion
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Jira"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Jira
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Todoist"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Todoist
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                    </div>
                    <div
                      class="menu-spacer hideOnWidescreen"
                    />
                    <div
                      class="menu-options hideOnWidescreen"
                    >
                      <div
                        aria-label="Cancel"
                        class="MenuOption TextOption menu-option menu-cancel"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Cancel
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                class="MenuOption SubMenuOption menu-option open-left"
                id="lang"
              >
                <svg
                  class="SubmenuTriangleIcon Icon"
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="50,35 75,50 50,65"
                  />
                </svg>
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Set language
                </div>
              </div>
            </div>
            <div>
              <div
                aria-label="Random icons"
                class="MenuOption SwitchOption menu-option"
                role="button"
              >
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Random icons
                </div>
                <div
                  class="Switch override-main size--small on"
                >
                  <div
                    class="octo-switch-inner"
                  />
                </div>
              </div>
            </div>
            <div>
              <div
                aria-label="Product tour"
                class="MenuOption TextOption menu-option product-tour"
                role="button"
              >
                <div
                  class="d-flex"
                >
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="menu-name"
                >
                  Product tour
                </div>
                <div
                  class="noicon"
                />
              </div>
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="d-flex"
              >
                <div
                  class="noicon"
                />
              </div>
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/sidebar/GlobalHeaderSettingsMenu languages menu open should match snapshot 1`] = `
<div>
  <div
    class="GlobalHeaderSettingsMenu"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="GlobalHeaderComponent__button menu-entry"
      >
        <i
          class="CompassIcon icon-settings-outline SettingsIcon"
        />
      </div>
      <div
        class="Menu noselect left "
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div>
              <div
                class="MenuOption SubMenuOption menu-option open-left"
                id="import"
              >
                <svg
                  class="SubmenuTriangleIcon Icon"
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="50,35 75,50 50,65"
                  />
                </svg>
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Import
                </div>
              </div>
            </div>
            <div>
              <div
                class="MenuOption SubMenuOption menu-option open-left menu-option-active"
                id="lang"
              >
                <svg
                  class="SubmenuTriangleIcon Icon"
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="50,35 75,50 50,65"
                  />
                </svg>
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Set language
                </div>
                <div
                  class="SubMenu Menu noselect left-bottom"
                >
                  <div
                    class="menu-contents"
                  >
                    <div
                      class="menu-options"
                    >
                      <div
                        aria-label="English"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          English
                        </div>
                        <svg
                          class="CheckIcon Icon"
                          viewBox="0 0 100 100"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <polyline
                            points="20,60 40,80 80,40"
                          />
                        </svg>
                      </div>
                      <div
                        aria-label="Español"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Español
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Deutsch"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Deutsch
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="日本語"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          日本語
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Français"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Français
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Nederlands"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Nederlands
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Pусский"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Pусский
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="中文 (繁體)"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          中文 (繁體)
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="中文 (简体)"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          中文 (简体)
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Türkçe"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Türkçe
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Occitan"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Occitan
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Português (Brasil)"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Português (Brasil)
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Català"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Català
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Ελληνικά"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Ελληνικά
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="bahasa Indonesia"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          bahasa Indonesia
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Italiano"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Italiano
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                      <div
                        aria-label="Svenska"
                        class="MenuOption TextOption menu-option"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Svenska
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                    </div>
                    <div
                      class="menu-spacer hideOnWidescreen"
                    />
                    <div
                      class="menu-options hideOnWidescreen"
                    >
                      <div
                        aria-label="Cancel"
                        class="MenuOption TextOption menu-option menu-cancel"
                        role="button"
                      >
                        <div
                          class="d-flex"
                        >
                          <div
                            class="noicon"
                          />
                        </div>
                        <div
                          class="menu-name"
                        >
                          Cancel
                        </div>
                        <div
                          class="noicon"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                aria-label="Random icons"
                class="MenuOption SwitchOption menu-option"
                role="button"
              >
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Random icons
                </div>
                <div
                  class="Switch override-main size--small on"
                >
                  <div
                    class="octo-switch-inner"
                  />
                </div>
              </div>
            </div>
            <div>
              <div
                aria-label="Product tour"
                class="MenuOption TextOption menu-option product-tour"
                role="button"
              >
                <div
                  class="d-flex"
                >
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="menu-name"
                >
                  Product tour
                </div>
                <div
                  class="noicon"
                />
              </div>
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="d-flex"
              >
                <div
                  class="noicon"
                />
              </div>
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/sidebar/GlobalHeaderSettingsMenu settings menu closed should match snapshot 1`] = `
<div>
  <div
    class="GlobalHeaderSettingsMenu"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="GlobalHeaderComponent__button menu-entry"
      >
        <i
          class="CompassIcon icon-settings-outline SettingsIcon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/sidebar/GlobalHeaderSettingsMenu settings menu open should match snapshot 1`] = `
<div>
  <div
    class="GlobalHeaderSettingsMenu"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="GlobalHeaderComponent__button menu-entry"
      >
        <i
          class="CompassIcon icon-settings-outline SettingsIcon"
        />
      </div>
      <div
        class="Menu noselect left "
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div>
              <div
                class="MenuOption SubMenuOption menu-option open-left"
                id="import"
              >
                <svg
                  class="SubmenuTriangleIcon Icon"
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="50,35 75,50 50,65"
                  />
                </svg>
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Import
                </div>
              </div>
            </div>
            <div>
              <div
                class="MenuOption SubMenuOption menu-option open-left"
                id="lang"
              >
                <svg
                  class="SubmenuTriangleIcon Icon"
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="50,35 75,50 50,65"
                  />
                </svg>
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Set language
                </div>
              </div>
            </div>
            <div>
              <div
                aria-label="Random icons"
                class="MenuOption SwitchOption menu-option"
                role="button"
              >
                <div
                  class="noicon"
                />
                <div
                  class="menu-name"
                >
                  Random icons
                </div>
                <div
                  class="Switch override-main size--small on"
                >
                  <div
                    class="octo-switch-inner"
                  />
                </div>
              </div>
            </div>
            <div>
              <div
                aria-label="Product tour"
                class="MenuOption TextOption menu-option product-tour"
                role="button"
              >
                <div
                  class="d-flex"
                >
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="menu-name"
                >
                  Product tour
                </div>
                <div
                  class="noicon"
                />
              </div>
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="d-flex"
              >
                <div
                  class="noicon"
                />
              </div>
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
